src/Streamly/Internal/Data/Array/Foreign/Mut/Types.hs

```diff
@@ -523,6 +524,33 @@ read = Unfold step inject
             return $ D.Yield x
                 (ReadUState fp (p `plusPtr` sizeOf (undefined :: a)))
 
+-- | Unfold an array into a stream in reverse order.
+--
+-- /Internal/
+{-# INLINE_NORMAL readRev #-}
+readRev :: forall m a. (Monad m, Storable a) => Unfold m (Array a) a
+readRev = Unfold step inject
+    where
+
+    inject (Array fp end _) =
+        let p = end `plusPtr` negate (sizeOf (undefined :: a))
+         in return $ ReadUState fp p
+
+    {-# INLINE_LATE step #-}
+    step (ReadUState fp@(ForeignPtr start _) p) | p < Ptr start =
+        let x = unsafeInlineIO $ touchForeignPtr fp
```

Comment:
How would this work? The `touchForeignPtr` is no longer sequential. It can be evaluated anytime according to what the compiler decides. We probably shouldn't `unsafeInlineIO` here.

Reply:
`touchForeignPtr fp >> return D.Stop` should work?

Reply:
1. There is a `seq` in the next line.
2. touchForeignPtr fp >> return D.Stop would require m to be IO Monad/MonadIO. It cannot work in Identity.

Reply:
`seq` does not enforce an ordering? `seq a b` says both `a` and `b` will be evaluated before `seq` returns anything.

Reply:
Our intention is not to enforce any ordering here. Our intention is to keep the pointer alive up to this point. It does not matter whether x is evaluated before or after. The only point is to not let the compiler optimize away x thinking that its unused.